### PR TITLE
Switch to swagger based typing and deserialization to typed map

### PR DESCRIPTION
**Pull request by dswiecki**

### CODE REVIEW

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/JsonSchemaTypeDefinitionExtractor.scala [hunk 1]
```diff
@@ -43,6 +46,19 @@ class JsonSchemaTypeDefinitionExtractor {
     }
   }
 
+  def parseCombinedSchema(s: CombinedSchema): TypingResult = s.getSubschemas.toList match {
+    case Nil => Typed.typedClass[Null] //???
+    case schema :: Nil => typeDefinition(schema)
+    //todo make it simple type schema only
+    case List(a: EnumSchema, b: StringSchema) => typeDefinition(b)
+    case List(b: StringSchema, a: EnumSchema) => typeDefinition(b)
+    case _ => throw new IllegalArgumentException(s"Schema '${s.toString}' is not supported yet.")
```
Comment (lciolecki) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/JsonSchemaTypeDefinitionExtractor.scala:55]: You don't need to run `.toString` method.

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/JsonSchemaTypeDefinitionExtractor.scala [hunk 2]
```diff
@@ -43,6 +46,19 @@ class JsonSchemaTypeDefinitionExtractor {
     }
   }
 
+  def parseCombinedSchema(s: CombinedSchema): TypingResult = s.getSubschemas.toList match {
+    case Nil => Typed.typedClass[Null] //???
+    case schema :: Nil => typeDefinition(schema)
+    //todo make it simple type schema only
+    case List(a: EnumSchema, b: StringSchema) => typeDefinition(b)
+    case List(b: StringSchema, a: EnumSchema) => typeDefinition(b)
```
Comment (lciolecki) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/JsonSchemaTypeDefinitionExtractor.scala:54]: a, b?

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala [hunk 3]
```diff
@@ -96,6 +107,15 @@ object SwaggerObject {
   }
 }
 
+object SwaggerJson {
```
Comment (mproch) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala:110]: is it used?

Comment (dswiecki) [reply utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala]: removed

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/serde/CirceJsonDeserializer.scala [hunk 4]
```diff
@@ -20,6 +21,6 @@ class CirceJsonDeserializer(jsonSchema: Schema) {
   def deserialize(string: String): Validated[circe.Error, Any] = {
     val jsonObject = new JSONTokener(string).nextValue()
     jsonSchema.validate(jsonObject)
-    Validated.fromEither(CirceUtil.decodeJson[Json](string).map(JsonToTypedMapConverter.jsonToAny))
+    Validated.fromEither(CirceUtil.decodeJson[Json](string).map(JsonToObject.apply(_, SwaggerBasedJsonSchemaTypeDefinitionExtractor.swaggerType(jsonSchema))))
```
Comment (mproch) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/serde/CirceJsonDeserializer.scala:24]: the type should be computed only once for this class?

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/JsonSchemaTypeDefinitionExtractor.scala [hunk 5]
```diff
@@ -1,10 +1,11 @@
...
 
-import org.everit.json.schema.{ArraySchema, BooleanSchema, FalseSchema, NullSchema, NumberSchema, ObjectSchema, Schema, StringSchema, TrueSchema}
+import org.everit.json.schema._
 import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypedObjectTypingResult, TypingResult}
 
 import java.time.{LocalDate, LocalDateTime, LocalTime}
 import scala.collection.JavaConverters._
+import scala.collection.convert.ImplicitConversions.`iterable AsScalaIterable`
 
 class JsonSchemaTypeDefinitionExtractor {
```
Comment (mproch) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/JsonSchemaTypeDefinitionExtractor.scala:10]: Are we using these changes?

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/SwaggerBasedJsonSchemaTypeDefinitionExtractor.scala [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+       |  "components": {
+       |    "schemas": {
+       |      "$fakeSchema": ${schema.toString}
+       |    }
+       |  }
+       |}""".stripMargin
+  }
+
+
+  def swaggerTypeDefinition(schema: Schema): TypingResult = {
```
Comment (mproch) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/SwaggerBasedJsonSchemaTypeDefinitionExtractor.scala:30]: do we need this method? Or we just invoke .typingResult on SwaggerTyped?

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala [hunk 7]
```diff
@@ -40,10 +42,11 @@ object SwaggerTyped {
     case objectSchema: ObjectSchema => SwaggerObject(objectSchema, swaggerRefSchemas)
     case mapSchema: MapSchema => SwaggerObject(mapSchema, swaggerRefSchemas)
     case arraySchema: ArraySchema => SwaggerArray(arraySchema, swaggerRefSchemas)
+    case objectSchema: JsonSchema if Option(objectSchema.getTypes).exists(_.contains("object")) => SwaggerObject(objectSchema, swaggerRefSchemas)
     case _ => Option(schema.get$ref()) match {
       case Some(ref) =>
         SwaggerTyped(swaggerRefSchemas(ref), swaggerRefSchemas)
-      case None => (schema.getType, Option(schema.getFormat)) match {
+      case None => (Option(schema.getType).getOrElse(schema.getTypes.asScala.head), Option(schema.getFormat)) match {
```
Comment (mproch) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala:49]: what about unions? If we don't want to handle them now, let's add TODO?

**Review by mproch — COMMENTED** [review]
Cool. added some minor comments

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/serde/CirceJsonDeserializer.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (gadomsky) [reply utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/serde/CirceJsonDeserializer.scala:24]: and also maybe it's worth now to optimize `ConfluentJsonSchemaPayloadDeserializer`  to use cache:

```scala
    val deserializer = readerSchemaData.schemaIdOpt.flatMap(
      id => schemaIdToCirceJsonDeserializerCache.get(id)
    ).getOrElse(new CirceJsonDeserializer(readerSchemaData.schema.rawSchema()))
   
```
WDYT?

[the remark this answers is not in the document]
Comment (dswiecki) [reply utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/serde/CirceJsonDeserializer.scala]: We'll fix caching in another PR

Comment (dswiecki) [reply utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/serde/CirceJsonDeserializer.scala]: done

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala [hunk 8]
```diff
@@ -79,6 +88,8 @@ object SwaggerTyped {
       Typed.typedClass[java.math.BigDecimal]
     case SwaggerDateTime =>
       Typed.typedClass[LocalDateTime]
+    case SwaggerNull =>
+      Typed.typedClass[Null]
```
Comment (mproch) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala:92]: Why not TypedNull?

Comment (dswiecki) [reply utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala]: done

**Review by mproch — APPROVED** [review]
Cool, thanks

**Pull request merged**